### PR TITLE
Fix interactive

### LINKED ISSUE #74: Error
Dear Kai,

I ran Karoo GP as usual:

sudo nice -1 python karoo-gp.py  

and the following error stopped Pause from loading at run-end:

Screen Shot 2022-06-26 at 11 42 13 AM

### PR DESCRIPTION
**Pull request by granawkins**
Fixed several broken features used in 'interactive' mode, i.e. when you run `python3 karoo-gp.py`. 

Still to do:
- Save/Load model and population (need to discuss)
- Use [click](https://pypi.org/project/click/) to format print statements

Also: This should fix #74 and #75

### CODE REVIEW

File: karoo_gp/pause.py [hunk 1]
```diff
@@ -233,66 +233,44 @@ def pause(menu_dict):
...
+                    break
+                elif int(query) in range(1, menu_dict[pop_len_key] + 1):
+                    menu_dict['input_a'] = 'eval'
+                    menu_dict['input_b'] = int(query)
+                    break
+                else:
+                    raise ValueError()
+            except ValueError:
+                print('\n\t\033[32m Enter a number from 1 including %s. '
+                        'Try again ...\033[0;0m' % str(menu_dict[pop_len_key]))
```
Comment (granawkins) [inline karoo_gp/pause.py:251]:
```suggestion
                        'Try again ...\033[0;0m' % str(menu_dict[pop_len_key] + 1))
```

File: karoo-gp.py [hunk 2]
```diff
@@ -483,8 +483,10 @@ def fx_karoo_pause(model):
         return 2
 
     elif input_a == 'eval':  # evaluate a Tree against the TEST data
-        # Display tree info
-        tree = model.population.next_gen_trees[input_b - 1]
+        if menu_dict['next_gen_len'] > 0:
+            tree = model.population.next_gen_trees[input_b - 1]
+        else:
+            tree = model.population.trees[input_b - 1]
```
Comment (ezio-melotti) [inline karoo-gp.py:489]: Should there be any indication that is using the current vs next gen population?

File: karoo_gp/branch.py [hunk 3]
```diff
@@ -223,7 +223,20 @@ def save(self):
...
+        if method == 'list':
+            return self.display_list(*args, **kwargs)
+        elif method == 'viz':
+            return self.display_viz(*args, **kwargs)
+
+    def display_list(self, prefix=''):
+        output = prefix + repr(self.node) + '\n'
+        if self.children:
+            for child in self.children:
+                output += child.display_list(prefix = prefix + '   ')
```
Comment (ezio-melotti) [inline karoo_gp/branch.py:236]:
```suggestion
                output += child.display_list(prefix=prefix+'   ')
```
Also using `join` is better here.

File: karoo_gp/pause.py [hunk 4]
```diff
@@ -233,66 +233,44 @@ def pause(menu_dict):
...
-                    print('\n\n\t\033[32m Enter \033[1mq\033[0;0m\033[32m to quit\033[0;0m')
+        has_next_gen = menu_dict['next_gen_len'] > 0
+        menu_dict['input_a'] = 'print_b' if has_next_gen else 'print_a'
+        pop_len_key = 'next_gen_len' if has_next_gen else 'population_len'
+        while True:
+            try:
+                query = input('\n\t Select a Tree to print: ')
+                if query == '':
+                    break
+                elif (int(query) in list(range(1, menu_dict[pop_len_key] + 1))):
```
Comment (ezio-melotti) [inline karoo_gp/pause.py:264]:
```suggestion
                elif int(query) in range(1, menu_dict[pop_len_key] + 1):
```

File: karoo_gp/branch.py (hunk 3 above)
Comment (granawkins) [inline karoo_gp/branch.py:236]:
```suggestion
                output = ''.join([output, child.display_list(prefix=prefix+'   ')])
```

File: karoo-gp.py (hunk 2 above)
Comment (granawkins) [reply karoo-gp.py:489]: That would be good - for now I'm just mimicking the old behavior. There's a lot of duplicated code as well, checking whether to use the `trees` or `next_gen_trees`. I will add a comment for now and lets discuss.
```suggestion
            # TODO: Tell user which population is being used
            tree = model.population.trees[input_b - 1]
```

File: karoo_gp/branch.py [hunk 5]
```diff
@@ -223,7 +223,20 @@ def save(self):
...
+        if method == 'list':
+            return self.display_list(*args, **kwargs)
+        elif method == 'viz':
+            return self.display_viz(*args, **kwargs)
+
+    def display_list(self, prefix=''):
+        output = prefix + repr(self.node) + '\n'
+        if self.children:
+            output += (''.join(child.display_list(prefix=prefix+' ')
+                       for child in self.children))
```
Comment (granawkins) [inline karoo_gp/branch.py:236]:
```suggestion
            output += ''.join(child.display_list(prefix=prefix+' ')
                              for child in self.children)
```

**Review by ezio-melotti — APPROVED** [review]

**Pull request merged**